index.php

```diff
@@ -27,6 +27,75 @@ function gutenberg_menu() {
 }
 add_action( 'admin_menu', 'gutenberg_menu' );
 
+
+$registered_blocks = array();
+
+/**
+ * Registers a block.
+ *
+ * @param  string   $name   Block name including namespace.
+ * @param  function $render Render function called to generate the HTML output
+ */
+function register_block( $name, $render ) {
+	global $registered_blocks;
+	$registered_blocks[ $name ] = $render;
+}
+
+/**
+ * Extract the block attributes from the block's attributes string
+ *
+ * @since 0.1.0
+ *
+ * @param  string $attr_string Attributes string
+
+ * @return array
+ */
+function parse_block_attributes( $attr_string ) {
+	$attributes_matcher = '/([^\s]+):([^\s]+)\s*/';
+	preg_match_all( $attributes_matcher, $attr_string, $matches );
+	$attributes = array();
+	foreach ( $matches[ 1 ] as $index => $attribute_match ) {
+		$attributes[ $attribute_match ] = $matches[ 2 ][ $index ];
+	}
+
+	return $attributes;
+}
+
+/**
+ * Renders the dynamic blocks into the post content
+ *
+ * @since 0.1.0
+ *
+ * @param  string $content Post content
+
+ * @return string          Updated post content
+ */
+function do_dynamic_blocks( $content ) {
+	global $registered_blocks;
+	$open_matcher = '/<!--\s*wp:([a-z](?:[a-z0-9\/]+)*)\s+((?:(?!-->).)*)-->.*<!--\s*\/wp:\g1\s+-->/';
```

Comment:
We shouldn't prescribe what save returns, because it may make sense to return a link, or a short message. Consider a block that renders the last 5 recent posts from a certain category. It may be good that when the dynamic markup can't be generated, it returns a link to `/blog/category/:term` with some text saying "View latest posts on :term".